common/src/main/java/com/microsoft/identity/common/internal/result/ResultFuture.java

```diff
@@ -28,21 +59,55 @@ public boolean isDone() {
     @Override
     public T get() throws InterruptedException {
         mCountDownLatch.await();
+
+        if (null != mException) {
+            throw new RuntimeException(mException);
+        }
+
         return mResult;
     }
 
     @Override
-    public T get(long l, TimeUnit timeUnit) throws InterruptedException, TimeoutException {
+    public T get(final long l, @NonNull final TimeUnit timeUnit) throws InterruptedException, TimeoutException {
         if (mCountDownLatch.await(l, timeUnit)) {
+            if (null != mException) {
+                throw new RuntimeException(mException);
+            }
+
             return mResult;
         } else {
-            throw new TimeoutException();
+            throw new TimeoutException(
+                    "Timed out waiting for: "
+                            + l // duration
+                            + timeUnit.name() // units
+            );
         }
+    }
+
+    public synchronized void setException(@NonNull final Exception exception) {
+        mException = exception;
+        mCountDownLatch.countDown();
 
+        for (final BiConsumer<T, Throwable> consumer : mConsumers) {
+            consumer.accept(mResult, exception);
+        }
     }
 
-    public void setResult(T result) {
+    public synchronized void setResult(@Nullable final T result) {
         mResult = result;
         mCountDownLatch.countDown();
+
+        for (final BiConsumer<T, Throwable> consumer : mConsumers) {
+            consumer.accept(result, mException);
+        }
+    }
```

Comment:
At this point, you could drop all references in this list.

Reply:
47d8fa9856e1f784755abe10c2a61311ed78712d